### DISCUSSION TITLE
Change LatLon of Point with Zoom

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#870] [Q&A] [answered]
By default, point text labels have fixed LatLon coordinates which are constant across zoom levels. However, the ideal placement can depend on the zoom level.

I would like to move labels of countries depending on the zoom level. This requires that I can configure the LatLon of a Point depending on the zoom level. Are there any options to do this in Planetiler?

**Comment by msbarry** [comment]
There isn't a built in way to do this, but you could emit distinct separate features for each zoom level. Something like:

```java 
for (int z = 1; z <= 14; z++) {
  var geom = computePointAtZoom(z);
  features.geometry(layer, geom).setZoomRange(z)... setup feature
}
```

Maybe there could be some more succinct way using the z or m coordinate of jts point features, but this approach should work today.

**Reply by wipfli** [reply]
Thanks for the suggestion. I might have a something wrong in the following code snippet since it does not produce a point feature in the output tile. Do you see anything obvious?

```java
  @Override
  public void processFeature(SourceFeature sourceFeature, FeatureCollector features) {
      double longitude = 0.0;
      double latitude = 0.0;
      Coordinate coordinate = new Coordinate(longitude, latitude);
      Point point = GeoUtils.JTS_FACTORY.createPoint(coordinate);

      features.geometry("country", point)
        .setZoomRange(0, 15)
        .setAttr("hello", "world");
```

**Reply by msbarry** [reply, marked as answer]
The geometry needs to already be projected into "world web mercator" coordinates where 0,0=top left, and 1,1=bottom right. There's a `latLonToWorldCoords(geometry)` utility for this, so you could do:

```java
Point point = GeoUtils.latLonToWorldCoords(GeoUtils.JTS_FACTORY.createPoint(coordinate));
```

**Reply by wipfli** [reply]
That worked. Thank you!

![image](https://github.com/onthegomap/planetiler/assets/53421382/2f8f18d2-030b-468e-a2f3-60aa502dc994)

**Comment by 1ec5** [comment]
This sounds like something that a MapLibre stylesheet would have better control over, especially if the map can be rotated and/or tilted or if the font can vary.

**Reply by wipfli** [reply]
I can't argue with that for variable fonts and map orientation. But if you know the font then you can optimize for a default orientation. It is actually quite fun to move labels around by hand. Maybe I can share a demo at some point...
